pkg/nsx/services/securitypolicy/store.go

```diff
@@ -1,153 +1,132 @@
-			if err != nil {
-				for _, e := range err {
-					fatalErrors <- e
-				}
-			}
-			c, _ := a.(model.SecurityPolicy)
-			err2 := service.SecurityPolicyStore.Add(c)
-			if err2 != nil {
-				fatalErrors <- err2
-			}
-		}
-		cursor = response.Cursor
-		if cursor == nil {
-			break
-		}
-		c, _ := strconv.Atoi(*cursor)
-		if int64(c) >= *response.ResultCount {
-			break
+func InitializeStore(service *SecurityPolicyService) {
+	service.Store = make(map[string]cache.Indexer)
+	service.Store[ResourceTypeSecurityPolicy] = cache.NewIndexer(keyFuncSecurityPolicy, cache.Indexers{key: indexFuncSecurityPolicy})
+	service.Store[ResourceTypeGroup] = cache.NewIndexer(keyFuncGroup, cache.Indexers{key: indexFuncGroup})
+	service.Store[ResourceTypeRule] = cache.NewIndexer(keyFuncRule, cache.Indexers{key: indexFuncRule})
+}
+
+type OperateSecurityPolicyStore struct {
+	securityService *SecurityPolicyService
+}
+
+type OperateRuleStore struct {
+	securityService *SecurityPolicyService
+}
+
+type OperateGroupStore struct {
+	securityService *SecurityPolicyService
+}
+
+func (o *OperateSecurityPolicyStore) TransResourceToStore(entity *data.StructValue) error {
+	obj, err := Converter.ConvertToGolang(entity, model.SecurityPolicyBindingType())
+	if err != nil {
+		for _, e := range err {
+			return e
 		}
 	}
+	sp, _ := obj.(model.SecurityPolicy)
+	err2 := o.securityService.Store[ResourceTypeSecurityPolicy].Add(sp)
+	if err2 != nil {
+		return err2
+	}
+	return nil
 }
 
-func queryRule(service *SecurityPolicyService, wg *sync.WaitGroup, fatalErrors chan error) {
-	defer wg.Done()
-	queryParam := fmt.Sprintf("%s:%s", common.ResourceType, common.ResourceTypeRule) + " AND " + queryTagCondition(service)
-	var cursor *string = nil
-	pageSize := common.PageSize
-	for {
-		response, err := service.NSXClient.QueryClient.List(queryParam, cursor, nil, &pageSize, nil, nil)
-		err = common.TransError(err)
-		if _, ok := err.(nsxutil.PageMaxError); ok == true {
-			common.DecrementPageSize(&pageSize)
-			continue
+func (o *OperateRuleStore) TransResourceToStore(entity *data.StructValue) error {
+	obj, err := Converter.ConvertToGolang(entity, model.RuleBindingType())
+	if err != nil {
+		for _, e := range err {
+			return e
 		}
-		if err != nil {
-			fatalErrors <- err
-		}
-		for _, g := range response.Results {
-			a, err := Converter.ConvertToGolang(g, model.RuleBindingType())
-			if err != nil {
-				for _, e := range err {
-					fatalErrors <- e
-				}
-			}
-			c, _ := a.(model.Rule)
-			err2 := service.RuleStore.Add(c)
-			if err2 != nil {
-				fatalErrors <- err2
-			}
-		}
-		cursor = response.Cursor
-		if cursor == nil {
-			break
-		}
-		c, _ := strconv.Atoi(*cursor)
-		if int64(c) >= *response.ResultCount {
-			break
+	}
+	rule, _ := obj.(model.Rule)
+	err2 := o.securityService.Store[ResourceTypeRule].Add(rule)
+	if err2 != nil {
+		return err2
+	}
+	return nil
+}
+
+func (o *OperateGroupStore) TransResourceToStore(entity *data.StructValue) error {
```

Comment:
all the TransResourceToStore has the same code, just BindingType and dataType are different. Is it possible to provide a more generic function like "func TransResourceToStore(store, entity, bindingType, dataType)" ?

Reply:
ok